Activities/Database/UiPath.Database/DatabaseConnection.cs

```diff
@@ -94,6 +101,214 @@ public virtual int InsertDataTable(string tableName, DataTable dataTable)
+
+            var ds = new DataSet();
+            dbDA.Fill(ds);
+            ValidateTableStructure(tableName, dataTable, ds);
+
+            // Perform an initial count on the destination table.
+            SqlCommand commandRowCount = new SqlCommand(string.Format("SELECT COUNT(*) FROM {0}", tableName), (SqlConnection)_connection);
+            long countStart = Convert.ToInt32(commandRowCount.ExecuteScalar());
+
+            bulkOps.WriteToServer(dataTable);
+
+            // Perform a final count on the destination
+            // table to see how many rows were added.
+            long countEnd = Convert.ToInt32(commandRowCount.ExecuteScalar());
+
+            affectedRecords = countEnd - countStart;
+            return affectedRecords;
+        }
+
+        private long BulkInsertOracleManaged(IBulkOperations bulkOps, string tableName, DataTable dataTable, string connection, DbDataAdapter dbDA, IExecutorRuntime executorRuntime)
+        {
+            long affectedRecords;
+            //if Oracle
+
+            dbDA.SelectCommand = _connection.CreateCommand();
+            dbDA.SelectCommand.Transaction = _transaction;
+            dbDA.SelectCommand.CommandType = CommandType.Text;
+            dbDA.SelectCommand.CommandText = string.Format("select * from {0} where rownum = 1", tableName);
+
+            var ds = new DataSet();
+            dbDA.Fill(ds);
+            ValidateTableStructure(tableName, dataTable, ds);
+
+            // Perform an initial count on the destination table.
+            DbCommand commandRowCount = _connection.CreateCommand();
+            commandRowCount.Connection = _connection;
+            commandRowCount.CommandText = string.Format("SELECT COUNT(*) FROM {0}", tableName);
+
+            long countStart = Convert.ToInt32(commandRowCount.ExecuteScalar());
+
+            List<string> oracleList = FindAssembliesInGAC("oracle.manageddataaccess*");
+
+            if (oracleList.Count > 0)
+            {
+                bool bulkCopyIsPresent = false;
+                //try all drivers with latest version first
+                foreach (var item in oracleList.OrderByDescending(x => x))
+                {
+                    var oracle = Assembly.LoadFile(item);
+                    var bulkCopyType = oracle.GetType("Oracle.ManagedDataAccess.Client.OracleBulkCopy");
+
+                    if (bulkCopyType != null)
+                    {
+                        bulkOps.BulkCopyType = bulkCopyType;
+                        bulkOps.WriteToServer(dataTable);
+                        bulkCopyIsPresent = true;
+                    }
+                    else continue;
+                }
+
+                if (!bulkCopyIsPresent)
+                {
+                    //if no bulk insert possible, fallback to insert data table
+                    if (executorRuntime != null)
+                    {
+                        var message = new LogMessage
```

Comment:
extract in a separate function - actually 2 of the. first one checks executorRuntime.HasFeature( and calls the second one that performs the logging

Reply:
already separate method